go/libraries/doltcore/sqle/stats/stats_provider.go

```diff
@@ -142,48 +172,55 @@ func (s DoltHistogram) toSql() []*stats.Bucket {
 }
 
 type indexMeta struct {
-	db    string
-	table string
-	index string
-	cols  []string
+	qual         sql.StatQualifier
+	cols         []string
+	updateChunks []tree.Node
+	// [start, stop] ordinals for each chunk for update
+	updateOrdinals [][]uint64
+	preexisting    []DoltBucket
+	allAddrs       []hash.Hash
 }
 
 func NewProvider() *Provider {
 	return &Provider{
-		mu:      &sync.Mutex{},
-		dbStats: make(map[string]*dbStats),
+		mu:                &sync.Mutex{},
+		dbStats:           make(map[string]*dbStats),
+		autoRefreshCancel: make(map[string]context.CancelFunc),
 	}
 }
 
 // Provider is the engine interface for reading and writing index statistics.
 // Each database has its own statistics table that all tables/indexes in a db
 // share.
 type Provider struct {
-	mu             *sync.Mutex
-	latestRootAddr hash.Hash
-	dbStats        map[string]*dbStats
+	mu                *sync.Mutex
+	latestRootAddr    hash.Hash
+	dbStats           map[string]*dbStats
+	autoRefreshCancel map[string]context.CancelFunc
 }
 
 // each database has one statistics table that is a collection of the
 // table stats in the database
 type dbStats struct {
-	db         string
-	active     map[hash.Hash]int
-	stats      map[sql.StatQualifier]*DoltStats
-	currentMap prolly.Map
+	db                string
+	stats             map[sql.StatQualifier]*DoltStats
+	currentMap        prolly.Map
+	latestRoot        *doltdb.RootValue
+	latestTableHashes map[string]hash.Hash
+}
+
+func newDbStats(dbName string) *dbStats {
+	return &dbStats{db: dbName, stats: make(map[sql.StatQualifier]*DoltStats), latestTableHashes: make(map[string]hash.Hash)}
 }
 
 var _ sql.StatsProvider = (*Provider)(nil)
 
 // Init scans the statistics tables, populating the |stats| attribute.
 // Statistics are not available for reading until we've finished loading.
 func (p *Provider) Load(ctx *sql.Context, dbs []dsess.SqlDatabase) error {
-	p.mu.Lock()
```

Comment:
Is it okay that this doesn't have a lock anymore?

Reply:
rewrote how the concurrency control works, it's fairly straightforward but i might need to add some go tests for race detection